dialogue-target/src/main/java/com/palantir/dialogue/Request.java

```diff
@@ -249,81 +243,81 @@ private Builder() {}
         }
 
         private ListMultimap<String, String> mutableHeaderParams() {
-            if (headerParams == null) {
-                headerParams = MultimapBuilder.treeKeys(String.CASE_INSENSITIVE_ORDER)
-                        .arrayListValues()
-                        .build();
-
-                if (existingUnmodifiableHeaderParams != null) {
-                    existingUnmodifiableHeaderParams.forEach(headerParams::put);
-                    existingUnmodifiableHeaderParams = null;
+            if (!isHeaderMutable()) {
+                setHeaderMutable();
+                ListMultimap<String, String> mutable =
+                        Multimaps.newListMultimap(new TreeMap<>(String.CASE_INSENSITIVE_ORDER), MAP_VALUE_FACTORY);
+                if (!headerParams.isEmpty()) {
+                    // Outperforms mutable.putAll(headerParams)
+                    headerParams.forEach(mutable::put);
                 }
+                headerParams = mutable;
             }
             return headerParams;
         }
 
-        private ImmutableListMultimap.Builder<String, String> mutableQueryParams() {
-            if (queryParams == null) {
-                queryParams = ImmutableListMultimap.builder();
-
-                if (existingQueryParams != null) {
-                    existingQueryParams.forEach(queryParams::put);
-                    existingQueryParams = null;
-                }
+        private ListMultimap<String, String> mutableQueryParams() {
+            if (!isQueryMutable()) {
+                setQueryMutable();
+                queryParams = ArrayListMultimap.create(queryParams);
             }
             return queryParams;
         }
 
-        private ImmutableMap.Builder<String, String> mutablePathParams() {
-            if (pathParams == null) {
-                pathParams = ImmutableMap.builder();
-
-                if (existingPathParams != null) {
-                    existingPathParams.forEach(pathParams::put);
-                    existingPathParams = null;
-                }
+        private Map<String, String> mutablePathParams() {
+            if (!isPathMutable()) {
+                setPathMutable();
+                pathParams = new HashMap<>(pathParams);
             }
             return pathParams;
         }
 
         private ListMultimap<String, String> unmodifiableHeaderParams() {
-            if (existingUnmodifiableHeaderParams != null) {
-                return existingUnmodifiableHeaderParams;
-            }
+            return isHeaderMutable() ? Multimaps.unmodifiableListMultimap(headerParams) : headerParams;
```

Comment:
fwiw, I found this unmodifiable wrapper degrades the `forEach` performance pretty hard, because `AbstractMapBasedMultimap#forEach` has got a very natural implementation that just zooms through its inner map:
```java
  @Override
  public void forEach(BiConsumer<? super K, ? super V> action) {
    checkNotNull(action);
    map.forEach(
        (key, valueCollection) -> valueCollection.forEach(value -> action.accept(key, value)));
  }
```

Unfortunately `UnmodifiableListMultimap` drops through to the `default` implementation of `Multimap#forEach` which goes through the whole indirection of building up an entryset and then iterating through that 
```java
  default void forEach(BiConsumer<? super K, ? super V> action) {
    checkNotNull(action);
    entries().forEach(entry -> action.accept(entry.getKey(), entry.getValue()));
  }
```

Reply:
Given that no methods in Request actually mutate these maps, maybe we can just be brave and omit this?

Reply:
Let's fix this upstream: https://github.com/google/guava/issues/3911

Reply:
Hmm, I'd prefer not to remove the wrappers in this change, not sure it's the best idea long term. I wouldn't mind fully-mutable request objects to be honest, our servers use modifiable request objects and it hasn't really come back to bite us. Realistically we have  a pretty strong thread model for dialogue requests, after a request is created it's handed off to the next channel which takes over ownership... probably out of scope for this change but worth considering.